src/t8_schemes/t8_quads_w_rectangular_subelements/t8_subelements/t8_subelements_quad_cxx.cxx

```diff
@@ -0,0 +1,1523 @@
+  /* 3.3) determine, whether the subelement is the first or the second subelement at the face */
+  if (sub_id + 1 == cum_neigh_array[face_number] && split == 1) {
+    sub_face_id = 1;            /* second subelement */
+  }
+  else {
+    sub_face_id = 0;            /* first subelement */
+  }
+
+  location[0] = face_number;
+  location[1] = split;
+  location[2] = sub_face_id;
+}
+
+void
+t8_default_scheme_sub_c::t8_element_reset_subelement_values (t8_element *
+                                                             elem)
+{
+  t8_quad_with_subelements *pquad_w_sub = (t8_quad_with_subelements *) elem;
+
+  pquad_w_sub->dummy_is_subelement = T8_IS_NO_SUBELEMENT;
+  pquad_w_sub->subelement_type = T8_IS_NO_SUBELEMENT;
+  pquad_w_sub->subelement_id = T8_IS_NO_SUBELEMENT;
+}
+
+void
+t8_default_scheme_sub_c::t8_element_copy_subelement_values (const
+                                                            t8_element *
+                                                            source,
+                                                            t8_element * dest)
+{
+  const t8_quad_with_subelements *pquad_w_sub_source =
+    (const t8_quad_with_subelements *) source;
+  t8_quad_with_subelements *pquad_w_sub_dest =
+    (t8_quad_with_subelements *) dest;
+
+  pquad_w_sub_dest->dummy_is_subelement =
+    pquad_w_sub_source->dummy_is_subelement;
+  pquad_w_sub_dest->subelement_type = pquad_w_sub_source->subelement_type;
+  pquad_w_sub_dest->subelement_id = pquad_w_sub_source->subelement_id;
+}
+
+t8_element_shape_t
+  t8_default_scheme_sub_c::t8_element_shape (const t8_element_t * elem)
+{
+  const t8_quad_with_subelements *pquad_w_sub =
+    (const t8_quad_with_subelements *) elem;
+
+  T8_ASSERT (t8_element_is_valid (elem));
+
+  if (pquad_w_sub->dummy_is_subelement == T8_IS_NO_SUBELEMENT) {
+    return T8_ECLASS_QUAD;
+  }
+  else {                        /* for quads, all subelements are triangles */
+    return T8_ECLASS_TRIANGLE;
+  }
+}
+
+void
+t8_default_scheme_sub_c::t8_element_new (int length, t8_element_t ** elem)
+{
+  /* allocate memory for a quad with subelements */
+  t8_default_scheme_common_c::t8_element_new (length, elem);
+
+  int                 i;
+  for (i = 0; i < length; i++) {
+    t8_quad_with_subelements *pquad_w_sub =
+      (t8_quad_with_subelements *) elem[i];
+    t8_element_init (1, elem[i], 0);
+    /* set dimension of quad to 2 */
+    T8_QUAD_SET_TDIM ((p4est_quadrant_t *) & pquad_w_sub->p4q, 2);
+  }
+}
+
+void
+t8_default_scheme_sub_c::t8_element_init (int length, t8_element_t * elem,
+                                          int new_called)
+{
+  t8_quad_with_subelements *pquad_w_sub = (t8_quad_with_subelements *) elem;
+
+  int                 i;
+
+  for (i = 0; i < length; i++) {
+    /* initalize subelement parameters */
+    pquad_w_sub[i].dummy_is_subelement = T8_IS_NO_SUBELEMENT;
+    pquad_w_sub[i].subelement_type = T8_IS_NO_SUBELEMENT;
+    pquad_w_sub[i].subelement_id = T8_IS_NO_SUBELEMENT;
```

Comment:
Would be better to call the new reset_subelement_values function here to avoid code duplication.

Reply:
done

Reply:
I just noticed that the init function is not only called for length = 1. Now, I am even more confused how this function works.